<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Playing with OmicIDX • OmicIDXR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Playing with OmicIDX">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">OmicIDXR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/omicidx_intro.html">Playing with OmicIDX</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Playing with OmicIDX</h1>
                        <h4 class="author">Sean Davis</h4>
            
            <h4 class="date">3/14/2019</h4>
      
      
      <div class="hidden name"><code>omicidx_intro.Rmd</code></div>

    </div>

    
    
<div id="introduction-to-the-omicidx-api" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction-to-the-omicidx-api" class="anchor"></a>Introduction to the OmicIDX API</h1>
<p>OmicIDX parses and then serves public genomics repository metadata. These metadata are growing quickly, updated often, and are now very large when taken as a whole. Because of the interrelated nature of the metadata and the myriad approaches and use cases that exist, including search, bulk download, and even data mining, we serve the data via a GraphQL endpoint.</p>
<p>Currently, OmicIDX contains the SRA and Biosample metadata sets. These overlap with each other, but SRA metadata contains deeper metadata than Biosample data on the same samples. Biosample, on the other hand, contains many more samples and currently includes metadata about a subset of NCBI GEO, all SRA samples, and some additional samples from projects like Genbank.</p>
</div>
<div id="graphql-for-accessing-omicidx-data" class="section level1">
<h1 class="hasAnchor">
<a href="#graphql-for-accessing-omicidx-data" class="anchor"></a>GraphQL for accessing OmicIDX data</h1>
<p><a href="https://graphql.org/">GraphQL</a> is a query language for APIs and a runtime for fulfilling those queries with existing data. GraphQL provides a complete and understandable description of the data in the API, gives clients the power to ask for exactly what they need and nothing more, makes it easier to evolve APIs over time, and enables powerful developer tools.</p>
<p>GraphQL has only a <em>single url</em>, called the endpoint, which allows access to all data in the API. GraphQL is also a <em>query language</em>. It is the GraphQL query that is submitted to the GraphQL endpoint that results in data being returned.</p>
<div id="what-is-a-graphql-query" class="section level2">
<h2 class="hasAnchor">
<a href="#what-is-a-graphql-query" class="anchor"></a>What is a GraphQL query?</h2>
<p>A GraphQL query looks a bit like JSON, except without quotes or commas. Here is an example GraphQL query for a fictitious GraphQL API.</p>
<pre><code>{ 
  allCharacters {
    name
  }
}</code></pre>
<p>If we had a server that contained Star Wars trivia, the response from the server might look like:</p>
<pre><code>{ "data": {
    "allCharacters": [
      { 
        "name":"Luke"
      },
      { 
        "name": "Darth"
      },
      ...
    ]
  }
}</code></pre>
<p>If we changed the query to:</p>
<pre><code>{ 
  allCharacters {
    name
    mass
  }
}</code></pre>
<p>the response would now look like:</p>
<pre><code>{ "data": {
    "allCharacters": [
      { 
        "name":"Luke",
        "mass": 80
      },
      { 
        "name": "Darth",
        "mass": 140
      },
      ...
    ]
  }
}</code></pre>
</div>
<div id="how-do-i-know-what-is-in-the-graphql-endpoint" class="section level2">
<h2 class="hasAnchor">
<a href="#how-do-i-know-what-is-in-the-graphql-endpoint" class="anchor"></a>How do I know what is in the GraphQL endpoint?</h2>
<p>The GraphQL <strong>schema</strong> describes the data model(s) contained in the GraphQL endpoint. GraphQL is strongly typed, has the concept of relationships between data types, and is self-documenting. In fact, one can use the GraphQL endpoint to discover what is in the endpoint. I will not go into the details right now, but this <em>introspection</em> capability makes possible some powerful tooling. One of the most ubiguitous is the so-called <strong>Graph<em>i</em>QL</strong> (note the <em>i</em> in the name) tool.</p>
</div>
<div id="exercise-1" class="section level2">
<h2 class="hasAnchor">
<a href="#exercise-1" class="anchor"></a>Exercise 1</h2>
<p>Navigate to <a href="http://graphql-omicidx.cancerdatasci.org/graphiql">Graph<em>i</em>QL</a> and follow along with the video (no sound).<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a></p>
<iframe width="560" height="315" src="https://www.youtube.com/embed/1Zg_Fbt56kc" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen>
</iframe>
</div>
</div>
<div id="querying-omicidx-programmatically" class="section level1">
<h1 class="hasAnchor">
<a href="#querying-omicidx-programmatically" class="anchor"></a>Querying OmicIDX programmatically</h1>
<p>GraphQL is quite easy to work with programmatically. All queries are made via a post request to the GraphQL endpoint. The POST request needs to be JSON encoded and must include the <code>"query"</code> key. A simple example of the JSON for a basic query might look like:</p>
<pre><code>{ "query": "{ heros { name weight } }"}</code></pre>
<p>Note that the query string (ie., <code>"{ heros { name weight } }"</code>) is just a string. It is not formated as JSON itself.</p>
<ul>
<li>Current OmicIDX GraphQL endpoint<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>: <a href="http://graphql-omicidx.cancerdatasci.org/graphql" class="uri">http://graphql-omicidx.cancerdatasci.org/graphql</a>
</li>
</ul>
<p>For example, let us get the first 500 SRA studies (500 is a limit to the number of results that we can retrieve in one go. The GraphQL might look like this:</p>
<pre><code>{
  allSraStudies {
    edges {
      node {
        accession
        title
        abstract
      }
    }
  }
}</code></pre>
<p>From exercise 1, you know that you can copy this query into the Graph<em>i</em>QL browser and get results. How about using <a href="https://curl.haxx.se/">curl</a>? Note that I am doing some gymnastics below to make this work in one command line. In practice, one would probably save the GraphQL query as a file (in json format) and then post that file<a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a>.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="ex">curl</span> --silent \</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">  -X POST \</a>
<a class="sourceLine" id="cb7-3" data-line-number="3">  -H <span class="st">"Content-Type: application/json"</span> \</a>
<a class="sourceLine" id="cb7-4" data-line-number="4">  http://graphql-omicidx.cancerdatasci.org/graphql \</a>
<a class="sourceLine" id="cb7-5" data-line-number="5">  --data @- <span class="op">&lt;&lt; EOF</span></a>
<a class="sourceLine" id="cb7-6" data-line-number="6">  </a>
<a class="sourceLine" id="cb7-7" data-line-number="7">  { "query":"{</a>
<a class="sourceLine" id="cb7-8" data-line-number="8">  allSraStudies(first: 1) {</a>
<a class="sourceLine" id="cb7-9" data-line-number="9">    edges {</a>
<a class="sourceLine" id="cb7-10" data-line-number="10">      node {</a>
<a class="sourceLine" id="cb7-11" data-line-number="11">        accession</a>
<a class="sourceLine" id="cb7-12" data-line-number="12">        title</a>
<a class="sourceLine" id="cb7-13" data-line-number="13">        abstract</a>
<a class="sourceLine" id="cb7-14" data-line-number="14">      }</a>
<a class="sourceLine" id="cb7-15" data-line-number="15">    }</a>
<a class="sourceLine" id="cb7-16" data-line-number="16">  }</a>
<a class="sourceLine" id="cb7-17" data-line-number="17">}"</a>
<a class="sourceLine" id="cb7-18" data-line-number="18">}</a>
<a class="sourceLine" id="cb7-19" data-line-number="19"> EOF</a></code></pre></div>
<pre><code>## {"errors":[{"expose":true,"statusCode":400,"status":400,"body":"    { \"query\":\"{  allSraStudies(first: 1) {    edges {      node {        accession        title        abstract      }    }  }}\"} EOF","type":"entity.parse.failed"}]}</code></pre>
<p>Note that the return value is simply JSON. All our normal tools for working with JSON are available to us to process and manipulate the results.</p>
</div>
<div id="using-omicidx-from-r" class="section level1">
<h1 class="hasAnchor">
<a href="#using-omicidx-from-r" class="anchor"></a>Using OmicIDX from R</h1>
<p>As a review, we know that:</p>
<ul>
<li>GraphQL queries look like JSON, but are not quite the same.</li>
<li>We can use the <a href="http://graphql-omicidx.cancerdatasci.org/graphiql">Graph<em>i</em>QL</a> graphical tool to help us write our query, including autocompleting fields, etc.</li>
<li>GraphQL queries are <code>POST</code>ed to the graphql endpoint and results are returned as JSON.</li>
</ul>
<p>In order to query from R, then, we need to:</p>
<ol style="list-style-type: decimal">
<li>Define our query.</li>
<li>
<code>POST</code> the query, encoded as a simple JSON data structure.</li>
<li>Deal with the resulting JSON data that comes back as a result of the post in #2.</li>
</ol>
<div id="defining-the-query" class="section level2">
<h2 class="hasAnchor">
<a href="#defining-the-query" class="anchor"></a>Defining the query</h2>
<p>Again, the easiest way to define the query is to experiment witht he <a href="http://graphql-omicidx.cancerdatasci.org/graphiql">Graph<em>i</em>QL</a> query tool. Once a query works as expected, the query can be reused from R.</p>
<p>In this case, we are going to continue with the query we worked with above:</p>
<pre><code>{
  allSraStudies(first: 1) {
    edges {
      node {
        accession
        title
        abstract
      }
    }
  }
}</code></pre>
</div>
<div id="performing-the-query" class="section level2">
<h2 class="hasAnchor">
<a href="#performing-the-query" class="anchor"></a>Performing the query</h2>
<p>The <code>CRANpkg('httr')</code> package performs http requests, including <code>POST</code> requests.</p>
<p>With our query defined, we need to prepare the POST body which will be sent to the server as JSON. In R, basicc lists are the information equivalent of JSON objects.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">post_body =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">query =</span> <span class="st">"</span></a>
<a class="sourceLine" id="cb10-2" data-line-number="2"><span class="st">{</span></a>
<a class="sourceLine" id="cb10-3" data-line-number="3"><span class="st">  allSraStudies(first: 20) {</span></a>
<a class="sourceLine" id="cb10-4" data-line-number="4"><span class="st">    edges {</span></a>
<a class="sourceLine" id="cb10-5" data-line-number="5"><span class="st">      node {</span></a>
<a class="sourceLine" id="cb10-6" data-line-number="6"><span class="st">        accession</span></a>
<a class="sourceLine" id="cb10-7" data-line-number="7"><span class="st">        title</span></a>
<a class="sourceLine" id="cb10-8" data-line-number="8"><span class="st">        sraExperimentsByStudyAccession {</span></a>
<a class="sourceLine" id="cb10-9" data-line-number="9"><span class="st">          totalCount</span></a>
<a class="sourceLine" id="cb10-10" data-line-number="10"><span class="st">        }</span></a>
<a class="sourceLine" id="cb10-11" data-line-number="11"><span class="st">      }</span></a>
<a class="sourceLine" id="cb10-12" data-line-number="12"><span class="st">    }</span></a>
<a class="sourceLine" id="cb10-13" data-line-number="13"><span class="st">  }</span></a>
<a class="sourceLine" id="cb10-14" data-line-number="14"><span class="st">}</span></a>
<a class="sourceLine" id="cb10-15" data-line-number="15"><span class="st">"</span>)</a></code></pre></div>
<p>The URL for the OmicIDX endpoint is: <code>http://graphql-omicidx.cancerdatasci.org/graphql</code>.<a href="#fn4" class="footnote-ref" id="fnref4"><sup>4</sup></a></p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">endpoint =<span class="st"> "http://graphql-omicidx.cancerdatasci.org/graphql"</span></a></code></pre></div>
<p>I am going to lead the reader a bit here and jump to a “handler” function that will convert the returned JSON into convenient R data structures.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">handler =<span class="st"> </span><span class="cf">function</span>(response) {</a>
<a class="sourceLine" id="cb12-2" data-line-number="2">  jsonlite<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/jsonlite/topics/fromJSON">fromJSON</a></span>(httr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/httr/topics/content">content</a></span>(response, <span class="dt">as=</span><span class="st">'text'</span>),<span class="dt">flatten =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb12-3" data-line-number="3">}</a></code></pre></div>
<p>Finally, we are ready to perform our query.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">resp =<span class="st"> </span>httr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/httr/topics/POST">POST</a></span>(endpoint, <span class="dt">body =</span> post_body, <span class="dt">encode =</span> <span class="st">'json'</span>)</a>
<a class="sourceLine" id="cb13-2" data-line-number="2">resp</a></code></pre></div>
<pre><code>## Response [http://graphql-omicidx.cancerdatasci.org/graphql]
##   Date: 2019-03-15 21:24
##   Status: 200
##   Content-Type: application/json; charset=utf-8
##   Size: 3.84 kB</code></pre>
<p>And finally, convert the response to convenient R data structure.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1">result =<span class="st"> </span><span class="kw">handler</span>(resp)</a>
<a class="sourceLine" id="cb15-2" data-line-number="2">knitr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/knitr/topics/kable">kable</a></span>(result<span class="op">$</span>data[[<span class="dv">1</span>]]<span class="op">$</span>edges)</a></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">node.accession</th>
<th align="left">node.title</th>
<th align="right">node.sraExperimentsByStudyAccession.totalCount</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">DRP000001</td>
<td align="left">Bacillus subtilis subsp. natto BEST195 genome sequencing project</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">DRP000002</td>
<td align="left">Model organism for prokaryotic cell differentiation and development</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">DRP000003</td>
<td align="left">Comprehensive identification and characterization of the nucleosome structure</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">DRP000004</td>
<td align="left">Comprehensive identification and characterization of the transcripts, their expression levels and sub-cellular localizations</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">DRP000005</td>
<td align="left">Comprehensive identification and characterization of the transcripts, their expression levels and sub-cellular localizations</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">DRP000006</td>
<td align="left">Comprehensive identification and characterization of the transcripts, their expression levels and sub-cellular localizations</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">DRP000007</td>
<td align="left">Comprehensive identification and characterization of the binding sites of polymerase II</td>
<td align="right">2</td>
</tr>
<tr class="even">
<td align="left">DRP000008</td>
<td align="left">Comprehensive identification and characterization of the binding sites of polymerase II</td>
<td align="right">2</td>
</tr>
<tr class="odd">
<td align="left">DRP000009</td>
<td align="left">Subsurface mine microbial mat metagenome</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">DRP000010</td>
<td align="left">Oryza sativa Japonica group genome sequencing project by QTL Genomics Research Center, Japan</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">DRP000011</td>
<td align="left">Comprehensive analysis of cytoplasmic mRNAs in HT29 cell.</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">DRP000012</td>
<td align="left">Comprehensive analysis of cytoplasmic mRNAs in HT29 cell at 4hr after treatment with tunicamycin.</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">DRP000013</td>
<td align="left">Comprehensive analysis of cytoplasmic mRNAs in HT29 cell at 16hr after treatment with tunicamycin.</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">DRP000014</td>
<td align="left">Comprehensive analysis of polysomal mRNAs in HT29 cell.</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">DRP000015</td>
<td align="left">Comprehensive analysis of polysomal mRNAs in HT29 cell at 4hr after treatment with tunicamycin.</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">DRP000016</td>
<td align="left">Comprehensive analysis of polysomal mRNAs in HT29 cell at 16hr after treatment with tunicamycin.</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">DRP000017</td>
<td align="left">Massive transcriptional start site mapping of Beas2B IL-4 stimulation cells.</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">DRP000018</td>
<td align="left">Massive transcriptional start site mapping of Beas2B IL-4 non-stimulation cells.</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">DRP000019</td>
<td align="left">Massive transcriptional start site mapping of Beas2B siRNA Stat6 IL-4 stimulation cells.</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">DRP000020</td>
<td align="left">Massive transcriptional start site mapping of Beas2B siRNA Stat6 IL-4 non-stimulation cells.</td>
<td align="right">1</td>
</tr>
</tbody>
</table>
<p>We can change our query a bit to get some more details.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1">study_experiment_counts_query =<span class="st"> "</span></a>
<a class="sourceLine" id="cb16-2" data-line-number="2"><span class="st">{</span></a>
<a class="sourceLine" id="cb16-3" data-line-number="3"><span class="st">  allSraStudies(first: 10) {</span></a>
<a class="sourceLine" id="cb16-4" data-line-number="4"><span class="st">    edges {</span></a>
<a class="sourceLine" id="cb16-5" data-line-number="5"><span class="st">      node {</span></a>
<a class="sourceLine" id="cb16-6" data-line-number="6"><span class="st">        accession</span></a>
<a class="sourceLine" id="cb16-7" data-line-number="7"><span class="st">        title</span></a>
<a class="sourceLine" id="cb16-8" data-line-number="8"><span class="st">        sraExperimentsByStudyAccession {</span></a>
<a class="sourceLine" id="cb16-9" data-line-number="9"><span class="st">          totalCount</span></a>
<a class="sourceLine" id="cb16-10" data-line-number="10"><span class="st">        }</span></a>
<a class="sourceLine" id="cb16-11" data-line-number="11"><span class="st">      }</span></a>
<a class="sourceLine" id="cb16-12" data-line-number="12"><span class="st">    }</span></a>
<a class="sourceLine" id="cb16-13" data-line-number="13"><span class="st">  }</span></a>
<a class="sourceLine" id="cb16-14" data-line-number="14"><span class="st">}</span></a>
<a class="sourceLine" id="cb16-15" data-line-number="15"><span class="st">"</span></a></code></pre></div>
<p>And streamline our code a bit as a function.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="co"># Borrow "handler" function from above</span></a>
<a class="sourceLine" id="cb17-2" data-line-number="2"><span class="co"># Borrow "endpoint" URL from above</span></a>
<a class="sourceLine" id="cb17-3" data-line-number="3">graphql_query =<span class="st"> </span><span class="cf">function</span>(gql) {</a>
<a class="sourceLine" id="cb17-4" data-line-number="4">  .handler =<span class="st"> </span><span class="cf">function</span>(response) {</a>
<a class="sourceLine" id="cb17-5" data-line-number="5">    jsonlite<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/jsonlite/topics/fromJSON">fromJSON</a></span>(httr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/httr/topics/content">content</a></span>(response, <span class="dt">as=</span><span class="st">'text'</span>),<span class="dt">flatten =</span> <span class="ot">TRUE</span>)<span class="op">$</span>data[[<span class="dv">1</span>]]<span class="op">$</span>edges</a>
<a class="sourceLine" id="cb17-6" data-line-number="6">  }</a>
<a class="sourceLine" id="cb17-7" data-line-number="7">  post_body =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">query =</span> gql)</a>
<a class="sourceLine" id="cb17-8" data-line-number="8">  resp =<span class="st"> </span>httr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/httr/topics/POST">POST</a></span>(endpoint, <span class="dt">body =</span> post_body, <span class="dt">encode =</span> <span class="st">'json'</span>)</a>
<a class="sourceLine" id="cb17-9" data-line-number="9">  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(<span class="kw">.handler</span>(resp))</a>
<a class="sourceLine" id="cb17-10" data-line-number="10">}</a></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1">knitr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/knitr/topics/kable">kable</a></span>(<span class="kw">graphql_query</span>(study_experiment_counts_query))</a></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">node.accession</th>
<th align="left">node.title</th>
<th align="right">node.sraExperimentsByStudyAccession.totalCount</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">DRP000001</td>
<td align="left">Bacillus subtilis subsp. natto BEST195 genome sequencing project</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">DRP000002</td>
<td align="left">Model organism for prokaryotic cell differentiation and development</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">DRP000003</td>
<td align="left">Comprehensive identification and characterization of the nucleosome structure</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">DRP000004</td>
<td align="left">Comprehensive identification and characterization of the transcripts, their expression levels and sub-cellular localizations</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">DRP000005</td>
<td align="left">Comprehensive identification and characterization of the transcripts, their expression levels and sub-cellular localizations</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">DRP000006</td>
<td align="left">Comprehensive identification and characterization of the transcripts, their expression levels and sub-cellular localizations</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">DRP000007</td>
<td align="left">Comprehensive identification and characterization of the binding sites of polymerase II</td>
<td align="right">2</td>
</tr>
<tr class="even">
<td align="left">DRP000008</td>
<td align="left">Comprehensive identification and characterization of the binding sites of polymerase II</td>
<td align="right">2</td>
</tr>
<tr class="odd">
<td align="left">DRP000009</td>
<td align="left">Subsurface mine microbial mat metagenome</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">DRP000010</td>
<td align="left">Oryza sativa Japonica group genome sequencing project by QTL Genomics Research Center, Japan</td>
<td align="right">1</td>
</tr>
</tbody>
</table>
<p>One might be interested in filtering the results, also. For example, to search study titles for “cancer” (case insensitive) we can add a filter to the query.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1">cancer_studies_query =<span class="st"> '</span></a>
<a class="sourceLine" id="cb19-2" data-line-number="2"><span class="st">{</span></a>
<a class="sourceLine" id="cb19-3" data-line-number="3"><span class="st">  allSraStudies(first: 10 filter: {title: {includesInsensitive: "cancer"}}) {</span></a>
<a class="sourceLine" id="cb19-4" data-line-number="4"><span class="st">    edges {</span></a>
<a class="sourceLine" id="cb19-5" data-line-number="5"><span class="st">      node {</span></a>
<a class="sourceLine" id="cb19-6" data-line-number="6"><span class="st">        accession</span></a>
<a class="sourceLine" id="cb19-7" data-line-number="7"><span class="st">        title</span></a>
<a class="sourceLine" id="cb19-8" data-line-number="8"><span class="st">      }</span></a>
<a class="sourceLine" id="cb19-9" data-line-number="9"><span class="st">    }</span></a>
<a class="sourceLine" id="cb19-10" data-line-number="10"><span class="st">  }</span></a>
<a class="sourceLine" id="cb19-11" data-line-number="11"><span class="st">}</span></a>
<a class="sourceLine" id="cb19-12" data-line-number="12"><span class="st">'</span></a>
<a class="sourceLine" id="cb19-13" data-line-number="13">knitr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/knitr/topics/kable">kable</a></span>(<span class="kw">graphql_query</span>(cancer_studies_query))</a></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">node.accession</th>
<th align="left">node.title</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">DRP000030</td>
<td align="left">human epigenomics sequencing project of breast cancer and normal cell lines</td>
</tr>
<tr class="even">
<td align="left">DRP000425</td>
<td align="left">Homo sapiens T24 human bladder cancer cell line Transcriptome</td>
</tr>
<tr class="odd">
<td align="left">DRP000628</td>
<td align="left">Identification and characterization of miRNA-mRNAs associations in a colon cancer cell line by massively paralleled sequencing</td>
</tr>
<tr class="even">
<td align="left">DRP000632</td>
<td align="left">Accumulation of genetic alterations in hepatic progenitor cells in mice lead to the development of liver cancers.</td>
</tr>
<tr class="odd">
<td align="left">DRP000989</td>
<td align="left">Biliary tract cancer targeted exome</td>
</tr>
<tr class="even">
<td align="left">DRP001071</td>
<td align="left">Quantitative detection of mutation alleles derived from lung cancer in plasma cell-free DNA by use of anomaly detection with deep sequencing data</td>
</tr>
<tr class="odd">
<td align="left">DRP001077</td>
<td align="left">Mutation profiles in inflamed gastric epithelium with Helicobacter pylori infection during the development of gastric cancer.</td>
</tr>
<tr class="even">
<td align="left">DRP001085</td>
<td align="left">The microRNA expression signature of bladder cancer assessed by deep sequencing</td>
</tr>
<tr class="odd">
<td align="left">DRP001358</td>
<td align="left">Single cell analysis of lung adenocarcinoma cell lines and the response to an anti-cancer drug stimulation</td>
</tr>
<tr class="even">
<td align="left">DRP002517</td>
<td align="left">Hypermutated human papillomavirus16 genome in cervical cancer</td>
</tr>
</tbody>
</table>
<p>To check the total count of available studies with “cancer” in the title, we can again write a query for that. Note that this query</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1">cancer_studies_query =<span class="st"> '</span></a>
<a class="sourceLine" id="cb20-2" data-line-number="2"><span class="st">{</span></a>
<a class="sourceLine" id="cb20-3" data-line-number="3"><span class="st">  allSraStudies(first: 10 filter: {title: {includesInsensitive: "cancer"}}) {</span></a>
<a class="sourceLine" id="cb20-4" data-line-number="4"><span class="st">    totalCount</span></a>
<a class="sourceLine" id="cb20-5" data-line-number="5"><span class="st">  }</span></a>
<a class="sourceLine" id="cb20-6" data-line-number="6"><span class="st">}</span></a>
<a class="sourceLine" id="cb20-7" data-line-number="7"><span class="st">'</span></a></code></pre></div>
<p>A keen eye will notice that there is not an <code>edges</code> component in this query. Our <code>handler</code> above returns the <code>edges</code> component, though. So, we are going to drop back to basic httr code here, again.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1">httr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/httr/topics/content">content</a></span>(</a>
<a class="sourceLine" id="cb21-2" data-line-number="2">  httr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/httr/topics/POST">POST</a></span>(endpoint, <span class="dt">body =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">query=</span>cancer_studies_query), <span class="dt">encode=</span><span class="st">'json'</span>),</a>
<a class="sourceLine" id="cb21-3" data-line-number="3">  <span class="dt">as=</span><span class="st">"parsed"</span></a>
<a class="sourceLine" id="cb21-4" data-line-number="4">)</a></code></pre></div>
<pre><code>## $data
## $data$allSraStudies
## $data$allSraStudies$totalCount
## [1] 2298</code></pre>
</div>
<div id="related-objects" class="section level2">
<h2 class="hasAnchor">
<a href="#related-objects" class="anchor"></a>Related objects</h2>
<p>One of the unique aspects of GraphQL is the ability to treat data as a graph, with data entities related to each other logically linked. For example, we can fetch SRA experiments and link them each to the study under which they were performed.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1">experiments_with_study =<span class="st"> '</span></a>
<a class="sourceLine" id="cb23-2" data-line-number="2"><span class="st">{</span></a>
<a class="sourceLine" id="cb23-3" data-line-number="3"><span class="st">  allSraExperiments(first: 5) {</span></a>
<a class="sourceLine" id="cb23-4" data-line-number="4"><span class="st">    edges{</span></a>
<a class="sourceLine" id="cb23-5" data-line-number="5"><span class="st">      node{</span></a>
<a class="sourceLine" id="cb23-6" data-line-number="6"><span class="st">        accession</span></a>
<a class="sourceLine" id="cb23-7" data-line-number="7"><span class="st">        title</span></a>
<a class="sourceLine" id="cb23-8" data-line-number="8"><span class="st">        libraryStrategy</span></a>
<a class="sourceLine" id="cb23-9" data-line-number="9"><span class="st">        sraStudyByStudyAccession{</span></a>
<a class="sourceLine" id="cb23-10" data-line-number="10"><span class="st">          title</span></a>
<a class="sourceLine" id="cb23-11" data-line-number="11"><span class="st">          accession</span></a>
<a class="sourceLine" id="cb23-12" data-line-number="12"><span class="st">        }</span></a>
<a class="sourceLine" id="cb23-13" data-line-number="13"><span class="st">      }</span></a>
<a class="sourceLine" id="cb23-14" data-line-number="14"><span class="st">    }</span></a>
<a class="sourceLine" id="cb23-15" data-line-number="15"><span class="st">  }</span></a>
<a class="sourceLine" id="cb23-16" data-line-number="16"><span class="st">}</span></a>
<a class="sourceLine" id="cb23-17" data-line-number="17"><span class="st">'</span></a></code></pre></div>
<p>The graphql query from R still works the same way, so we execute as before.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1">result =<span class="st"> </span><span class="kw">graphql_query</span>(experiments_with_study)</a>
<a class="sourceLine" id="cb24-2" data-line-number="2">knitr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/knitr/topics/kable">kable</a></span>(result)</a></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">node.accession</th>
<th align="left">node.title</th>
<th align="left">node.libraryStrategy</th>
<th align="left">node.sraStudyByStudyAccession.title</th>
<th align="left">node.sraStudyByStudyAccession.accession</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">DRX000001</td>
<td align="left">B. subtilis subsp. natto genome sequencing September 2008</td>
<td align="left">WGS</td>
<td align="left">Bacillus subtilis subsp. natto BEST195 genome sequencing project</td>
<td align="left">DRP000001</td>
</tr>
<tr class="even">
<td align="left">DRX000002</td>
<td align="left">B. subtilis subsp. subtilis genome resequencing September 2008</td>
<td align="left">WGS</td>
<td align="left">Model organism for prokaryotic cell differentiation and development</td>
<td align="left">DRP000002</td>
</tr>
<tr class="odd">
<td align="left">DRX000003</td>
<td align="left">DLD1_normoxia_nucleosome</td>
<td align="left">FL-cDNA</td>
<td align="left">Comprehensive identification and characterization of the nucleosome structure</td>
<td align="left">DRP000003</td>
</tr>
<tr class="even">
<td align="left">DRX000004</td>
<td align="left">DLD1_polysome</td>
<td align="left">FL-cDNA</td>
<td align="left">Comprehensive identification and characterization of the transcripts, their expression levels and sub-cellular localizations</td>
<td align="left">DRP000004</td>
</tr>
<tr class="odd">
<td align="left">DRX000005</td>
<td align="left">DLD1_cytoplasmic</td>
<td align="left">FL-cDNA</td>
<td align="left">Comprehensive identification and characterization of the transcripts, their expression levels and sub-cellular localizations</td>
<td align="left">DRP000005</td>
</tr>
</tbody>
</table>
</div>
<div id="fetching-all-results" class="section level2">
<h2 class="hasAnchor">
<a href="#fetching-all-results" class="anchor"></a>Fetching all results</h2>
<p><strong>TODO</strong></p>
<ul>
<li>Describe cursors</li>
</ul>
</div>
<div id="using-variables" class="section level2">
<h2 class="hasAnchor">
<a href="#using-variables" class="anchor"></a>Using variables</h2>
<p><strong>TODO</strong></p>
<ul>
<li>Describe adding variables to POST body</li>
</ul>
</div>
</div>
<div id="conclusion" class="section level1">
<h1 class="hasAnchor">
<a href="#conclusion" class="anchor"></a>Conclusion</h1>
<p><strong>TODO</strong></p>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>The urls in this document <em>are subject to change</em>.<a href="#fnref1" class="footnote-back">↩</a></p></li>
<li id="fn2"><p>The urls in this document <em>are subject to change</em>.<a href="#fnref2" class="footnote-back">↩</a></p></li>
<li id="fn3"><p>*For examples of how to post JSON using curl, see <a href="https://stackoverflow.com/questions/7172784/how-to-post-json-data-with-curl-from-terminal-commandline-to-test-spring-rest?rq=1">this stackoverflow post</a><a href="#fnref3" class="footnote-back">↩</a></p></li>
<li id="fn4"><p>The urls in this document <em>are subject to change</em>.<a href="#fnref4" class="footnote-back">↩</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction-to-the-omicidx-api">Introduction to the OmicIDX API</a></li>
      <li>
<a href="#graphql-for-accessing-omicidx-data">GraphQL for accessing OmicIDX data</a><ul class="nav nav-pills nav-stacked">
<li><a href="#what-is-a-graphql-query">What is a GraphQL query?</a></li>
      <li><a href="#how-do-i-know-what-is-in-the-graphql-endpoint">How do I know what is in the GraphQL endpoint?</a></li>
      <li><a href="#exercise-1">Exercise 1</a></li>
      </ul>
</li>
      <li><a href="#querying-omicidx-programmatically">Querying OmicIDX programmatically</a></li>
      <li>
<a href="#using-omicidx-from-r">Using OmicIDX from R</a><ul class="nav nav-pills nav-stacked">
<li><a href="#defining-the-query">Defining the query</a></li>
      <li><a href="#performing-the-query">Performing the query</a></li>
      <li><a href="#related-objects">Related objects</a></li>
      <li><a href="#fetching-all-results">Fetching all results</a></li>
      <li><a href="#using-variables">Using variables</a></li>
      </ul>
</li>
      <li><a href="#conclusion">Conclusion</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Sean Davis.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
